app/models/bus_stop.rb

```diff
@@ -23,7 +23,8 @@ class BusStop < ApplicationRecord
   def data_fields
     Field.categories.map do |category|
       category_fields = Field.in_category(category).order(:rank).map do |field|
-        BusStopField.where(field: field, bus_stop: self).first_or_create
+        # need to find last version of BusStopField or create it.
+        BusStopField.where(field: field, bus_stop: self).order(:created_at).last rescue BusStopField.create(field: field, bus_stop: self)
```

Comment:
No, that's definitely slower. We should use the built-in Rails 'find or whatever' methods rather than rescuing.

Also, will `self.bus_stop_fields.where(...)` work?

Reply:
Oh ok. It wasn't about speed. Specifically, the line you had was getting the wrong version. It was getting the oldest version of the `bus_stop_field`, and we want the most recent.

Reply:
I don't know why you're asking me the second question. And I don't know how to answer it.

Reply:
The wrong *version*? There shouldn't be multiple *versions* - we're not talking about PaperTrail. There should only be one BSF per field and stop. Don't we have a uniqueness validation in place for this?

I'm asking the second question because we have AR associations defined - may as well use them.

Reply:
OK I'll try to figure out what I mean via console. Sorry I used the term "version".

Reply:
```
BusStopField.where(field: field, bus_stop: self)
  BusStopField Load (0.8ms)  SELECT `bus_stop_fields`.* FROM `bus_stop_fields` WHERE `bus_stop_fields`.`field_name` = 'Accessible' AND `bus_stop_fields`.`bus_stop_id` = 12
=> [#<BusStopField:0x00007fd6b20afbf0
=> [#<BusStopField:0x00007fd6b20afbf0
  bus_stop_id: 12,
  value: nil,
  created_at: Tue, 10 Jul 2018 13:47:22 EDT -04:00,
  updated_at: Tue, 10 Jul 2018 13:47:22 EDT -04:00,
  field_name: "Accessible">,
 #<BusStopField:0x00007fd6b20af808
  bus_stop_id: 12,
  value: "0",
  created_at: Tue, 10 Jul 2018 15:08:13 EDT -04:00,
  updated_at: Tue, 10 Jul 2018 15:08:13 EDT -04:00,
  field_name: "Accessible">,
 #<BusStopField:0x00007fd6b20af178
  bus_stop_id: 12,
  value: "0",
  created_at: Tue, 10 Jul 2018 15:09:21 EDT -04:00,
  updated_at: Tue, 10 Jul 2018 15:09:21 EDT -04:00,
  field_name: "Accessible">,
 #<BusStopField:0x00007fd6b20aec78
  bus_stop_id: 12,
  value: "0",
  created_at: Tue, 10 Jul 2018 15:11:15 EDT -04:00,
  updated_at: Tue, 10 Jul 2018 15:11:15 EDT -04:00,
  field_name: "Accessible">,
 #<BusStopField:0x00007fd6b20ae570
  bus_stop_id: 12,
  value: "0",
  created_at: Tue, 10 Jul 2018 15:17:48 EDT -04:00,
  updated_at: Tue, 10 Jul 2018 15:17:48 EDT -04:00,
  field_name: "Accessible">,
 #<BusStopField:0x00007fd6b20adf30
  bus_stop_id: 12,
  value: "0",
  created_at: Tue, 10 Jul 2018 15:18:26 EDT -04:00,
  updated_at: Tue, 10 Jul 2018 15:18:26 EDT -04:00,
  field_name: "Accessible">,
 #<BusStopField:0x00007fd6b20ad800
  bus_stop_id: 12,
  value: "0",
  created_at: Tue, 10 Jul 2018 15:19:08 EDT -04:00,
  updated_at: Tue, 10 Jul 2018 15:19:08 EDT -04:00,
  field_name: "Accessible">,
```

Reply:
Well that's a problem. We should have a uniqueness validation in place on BSFs, which will hopefully reveal whatever code is creating duplicates.

Reply:
Yeah, I don't think we have a validation for that. I'll add it in and revert this line.

Reply:
Commented before I saw your comment

Reply:
Found it. It's when I call assign_attributes on `@stop`.

Reply:
Okay. This also answers https://github.com/umts/stop-project/pull/107#discussion_r201696027.